com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/linking/LazyLinkingResource2.java

```diff
@@ -134,7 +136,64 @@ public void setLoading(final boolean loading) {
     isLoading = loading;
   }
 
-  /** {@inheritDoc} */
+  /**
+   * Like the superclass implementation, but additionally handles any {@link StackOverflowError}s.
+   * Stack overflows can occur for deeply nested models.
+   * {@inheritDoc}
+   */
+  @Override
+  public void load(final Map<?, ?> options) throws IOException {
+    try {
+      super.load(options);
+    } catch (StackOverflowError e) {
+      trimStackOverflowErrorStackTrace(e);
+      LOGGER.warn("Failed to load " + uri + " from storage", e); //$NON-NLS-1$//$NON-NLS-2$
+
+      /*
+       * As for IOExceptions in the superclass implementation,
+       * clear and unload so that can fall back to reloading.
+       */
+      if (contents != null) {
+        contents.clear();
+      }
+      if (eAdapters != null) {
+        eAdapters.clear();
+      }
+      unload();
+
+      /*
+       * We actually want to call the super super class method.
+       * This should do something similar.
+       */
+      isLoading = true;
+      try {
+        super.load(options);
+      } finally {
+        isLoading = false;
+      }
+    }
+  }
+
+  /**
+   * Modify the argument StackOverflowError so that it only retains
+   * the first and last {@link #STACK_TRACE_LIMIT} stack trace elements of {@code ex}
+   * if it has more than that.
+   *
+   * @param ex
+   *          the StackOverflowError; may get modified
+   */
+  @SuppressWarnings("nls")
+  private static void trimStackOverflowErrorStackTrace(final StackOverflowError ex) {

```

Comment:
could we not refactor com.avaloq.tools.ddk.xtext.builder.MonitoredClusteringBuilderState.logStackOverflowErrorStackTrace(StackOverflowError, URI) and use it?
I would think that if we remove the URI parameter and do the logging outside as you do here, one could reuse the code. Shouldn't we do that?

Reply:
OK, I've moved the common exception trimming code into a utility method.